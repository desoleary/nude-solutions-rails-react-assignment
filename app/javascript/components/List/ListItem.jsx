import { List, Typography } from 'antd';
import { arrayOf, node, number, oneOfType, string } from 'prop-types';
import React from 'react';

const { Item } = List;
const { Text } = Typography;

Item.displayName = 'ItemAntd';

const ListItem = ({ text, extra, actions, id }) => {
  const key = `list-item-${id}`;

  return (
    <Item extra={extra} actions={actions} key={key}>
      <Text>{text}</Text>
    </Item>
  );
};

ListItem.defaultProps = {
  actions: []
};

ListItem.propTypes = {
  id: oneOfType([number, string]).isRequired,
  text: string,
  extra: node,
  actions: arrayOf(node)
};

export default ListItem;
