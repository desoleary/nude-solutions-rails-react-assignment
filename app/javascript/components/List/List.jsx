import { List as ListAntd } from 'antd';
import { arrayOf, node, number, oneOfType, shape, string } from 'prop-types';
import React from 'react';
import ListItem from './ListItem';

ListAntd.displayName = 'ListAntd';

const List = (props) => {
  const { dataSource } = props;

  return (
    <ListAntd
      dataSource={dataSource}
      size='small'
      split={false}
      renderItem={ListItem}
    />
  );
};

List.propTypes = {
  dataSource: arrayOf(
    shape({
      id: oneOfType([number, string]).isRequired,
      text: string,
      extra: node,
      actions: arrayOf(node)
    })
  )
};

export default List;
