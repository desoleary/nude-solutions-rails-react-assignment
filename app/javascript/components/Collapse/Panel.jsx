import { Collapse } from 'antd';
import { arrayOf, node, number, oneOfType, shape, string } from 'prop-types';
import React from 'react';
import List from '../List';

const { Panel: PanelAntd } = Collapse;
PanelAntd.displayName = 'PanelAntd';

const Panel = (props) => {
  const { id, header, extra, items, ...rest } = props;
  const key = `panel-${id}`;
  const listKey = `panel-list-${id}`;

  return (
    <PanelAntd {...rest} header={header} extra={extra} key={key}>
      <List dataSource={items} key={listKey} />
    </PanelAntd>
  );
};

Panel.propTypes = {
  id: oneOfType([string, number]),
  header: oneOfType([string, number, node]),
  extra: oneOfType([string, number, node]),
  items: arrayOf(
    shape({
      id: oneOfType([number, string]).isRequired,
      text: string,
      extra: node,
      actions: arrayOf(node)
    })
  )
};

export default Panel;
