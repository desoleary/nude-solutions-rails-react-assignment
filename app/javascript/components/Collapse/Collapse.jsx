import { Collapse as CollapseAntd } from 'antd';
import { noop } from 'lodash-es';
import {
  arrayOf,
  func,
  node,
  number,
  oneOfType,
  shape,
  string,
  bool
} from 'prop-types';
import React from 'react';
import Panel from './Panel';

CollapseAntd.displayName = 'CollapseAntd';

const Collapse = (props) => {
  const { data, footerProps, expandAllByDefault, ...rest } = props;

  const panelIdFormatter = (id) => `collapse-panel-${id}`;

  const defaultActiveKey = expandAllByDefault
    ? data.map(({ id }) => panelIdFormatter(id))
    : props.defaultActiveKey;

  return (
    <CollapseAntd
      {...rest}
      onChange={props.onChange}
      defaultActiveKey={defaultActiveKey}>
      {data.map((itemProps) => (
        <Panel {...itemProps} key={panelIdFormatter(itemProps.id)} />
      ))}
      {footerProps && (
        <Panel id='collapse-footer-panel' {...footerProps} disabled />
      )}
    </CollapseAntd>
  );
};

Collapse.defaultProps = {
  expandAllByDefault: false,
  onChange: noop
};

Collapse.propTypes = {
  data: arrayOf(
    shape({
      id: oneOfType([number, string]).isRequired,
      text: string,
      extra: node,
      actions: arrayOf(node)
    })
  ),
  expandAllByDefault: bool,
  defaultActiveKey: oneOfType([
    string,
    number,
    oneOfType([arrayOf(string), arrayOf(number)])
  ]),
  footerProps: shape({
    id: oneOfType([string, number]),
    header: oneOfType([string, number, node]),
    extra: oneOfType([string, number, node])
  }),
  onChange: func
};

export default Collapse;
