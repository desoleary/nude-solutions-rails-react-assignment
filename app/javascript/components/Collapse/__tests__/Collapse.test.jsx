import { Collapse as CollapseAntd } from 'antd';
import { mount, shallow } from 'enzyme';
import React from 'react';
import List from '../../List';

import Collapse from '../Collapse';

const props = {
  data: [
    {
      id: 'some-identifier',
      header: 'some-text',
      extra: <div>some-extra-content</div>
    }
  ],
  expandAllByDefault: true,
  footerProps: {
    id: 'some-footer-panel-identifier',
    header: 'some-footer-header',
    extra: <div>some-footer-extra-content</div>
  }
};

describe('Collapse', () => {
  let wrapper;

  context('with shallow rendering', () => {
    beforeEach(() => {
      wrapper = shallow(<Collapse {...props} />);
    });

    it('compares snapshot', () => {
      expect(wrapper.length).to.eql(1);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('with full rendering', () => {
    beforeEach(() => {
      wrapper = mount(<Collapse {...props} />);
    });

    it("passes props correctly to Antd's Panel", () => {
      const antdCollapseWrapper = wrapper.find(CollapseAntd);

      expect(antdCollapseWrapper.text()).to.include('some-text');
      expect(antdCollapseWrapper.text()).to.include('some-extra-content');
      expect(antdCollapseWrapper.text()).to.include('some-footer-header');
      expect(antdCollapseWrapper.text()).to.include(
        'some-footer-extra-content'
      );
    });

    it('passes props correctly to List', () => {
      const listWrapper = wrapper.find(List);
      const listProps = listWrapper.props();

      expect(listWrapper.key()).to.eql('panel-list-some-identifier');
      expect(listProps.dataSource).to.eql(props.items);
    });
  });
});
