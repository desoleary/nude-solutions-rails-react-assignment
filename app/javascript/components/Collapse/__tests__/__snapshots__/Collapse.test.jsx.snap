// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse with shallow rendering compares snapshot 1`] = `
<CollapseAntd
  bordered={true}
  defaultActiveKey={
    Array [
      "collapse-panel-some-identifier",
    ]
  }
  onChange={[Function]}
>
  <Panel
    extra={
      <div>
        some-extra-content
      </div>
    }
    header="some-text"
    id="some-identifier"
    key="collapse-panel-some-identifier"
  />
  <Panel
    disabled={true}
    extra={
      <div>
        some-footer-extra-content
      </div>
    }
    header="some-footer-header"
    id="some-footer-panel-identifier"
  />
</CollapseAntd>
`;
